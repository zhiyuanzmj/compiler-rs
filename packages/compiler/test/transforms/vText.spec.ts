import { DOMErrorCodes, NodeTypes } from '@vue/compiler-dom'
import { describe, expect, test, vi } from 'vitest'
import {
  IRNodeTypes,
  transformChildren,
  transformElement,
  transformVText,
} from '../../src'
import { makeCompile } from './_utils'

const compileWithVText = makeCompile({
  nodeTransforms: [transformElement, transformChildren],
  directiveTransforms: {
    text: transformVText,
  },
})

describe('v-text', () => {
  test('should convert v-text to setText', () => {
    const { code, ir, helpers } = compileWithVText(
      `<div v-text={str.value}></div>`,
    )

    expect(helpers).contains('setText')
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.GET_TEXT_CHILD,
        parent: 0,
      },
    ])

    expect(ir.block.effect).toMatchObject([
      {
        expressions: [
          {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: 'str.value',
            isStatic: false,
          },
        ],
        operations: [
          {
            type: IRNodeTypes.SET_TEXT,
            element: 0,
            values: [
              {
                type: NodeTypes.SIMPLE_EXPRESSION,
                content: 'str.value',
                isStatic: false,
              },
            ],
          },
        ],
      },
    ])

    expect(code).matchSnapshot()
  })

  test('should raise error and ignore children when v-text is present', () => {
    const onError = vi.fn()
    const { code, ir, templates } = compileWithVText(
      `<div v-text={test}>hello</div>`,
      {
        onError,
      },
    )
    expect(onError.mock.calls).toMatchObject([
      [{ code: DOMErrorCodes.X_V_TEXT_WITH_CHILDREN }],
    ])

    // children should have been removed
    expect(ir.templates).toEqual(['<div> </div>'])

    expect(ir.block.effect).toMatchObject([
      {
        expressions: [
          {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: 'test',
            isStatic: false,
          },
        ],
        operations: [
          {
            type: IRNodeTypes.SET_TEXT,
            element: 0,
            values: [
              {
                type: NodeTypes.SIMPLE_EXPRESSION,
                content: 'test',
                isStatic: false,
              },
            ],
          },
        ],
      },
    ])

    expect(code).matchSnapshot()
    // children should have been removed
    expect(templates).contains('_template("<div> </div>", true)')
  })

  test('should raise error if has no expression', () => {
    const onError = vi.fn()
    const { code } = compileWithVText(`<div v-text></div>`, { onError })
    expect(code).matchSnapshot()
    expect(onError.mock.calls).toMatchObject([
      [{ code: DOMErrorCodes.X_V_TEXT_NO_EXPRESSION }],
    ])
  })
})
