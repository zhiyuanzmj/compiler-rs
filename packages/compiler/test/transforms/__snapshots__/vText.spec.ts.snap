// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`v-text > should convert v-text to setText 1`] = `
"
  const n0 = t0()
  const x0 = _child(n0)
  _renderEffect(() => _setText(x0, _toDisplayString(str.value)))
  return n0
"
`;

exports[`v-text > should raise error and ignore children when v-text is present 1`] = `
"
  const n0 = t0()
  const x0 = _child(n0)
  _renderEffect(() => _setText(x0, _toDisplayString(test)))
  return n0
"
`;

exports[`v-text > should raise error if has no expression 1`] = `
"
  const n0 = t0()
  return n0
"
`;
